VMEncryption/main/EncryptionSettingsUtil.py

```diff
@@ -182,30 +186,47 @@ def _dict_to_name_value_array(self, values):
                     })
         return array
 
-    def get_settings_data(self, protector_name, kv_url, kv_id, kek_url, kek_kv_id, kek_algorithm, extra_device_items, disk_util, crypt_mount_config_util):
+    def get_settings_data(self, protector_name, kv_url, kv_id, kek_url, kek_kv_id, kek_algorithm, extra_device_items, disk_util, crypt_mount_config_util, key_store_type):
         """ returns encryption settings object in format required by wire server """
 
-        # validate key vault parameters prior to creating the encryption settings object
         cutil = CheckUtil(self.logger)
-        cutil.check_kv_id(kv_id, "A KeyVault ID is required, but is missing or invalid")
-        cutil.check_kv_url(kv_url, "A KeyVault URL is required, but is missing or invalid")
-        cutil.check_kv_name(kv_id, kv_url, "A KeyVault ID and KeyVault URL were provided, but their key vault names did not match")
-        if kek_url:
-            cutil.check_kv_id(kek_kv_id, "A KEK URL was specified, but its KEK KeyVault ID was missing or invalid")
-            cutil.check_kek_url(kek_url, "A KEK URL was specified, but it was invalid")
-            cutil.check_kek_name(kek_kv_id, kek_url, "A KEK ID and KEK URL were provided, but their key vault names did not match")
-            if kek_algorithm not in CommonVariables.encryption_algorithms:
-                if kek_algorithm:
-                    raise Exception("The KEK encryption algorithm requested was not recognized")
-                else:
-                    kek_algorithm = CommonVariables.default_encryption_algorithm
-                    self.logger.log("No KEK algorithm specified, defaulting to {0}".format(kek_algorithm))
+
+        # ManagedHSM Checks
+        if key_store_type.lower() == CommonVariables.KeyStoreTypeManagedHSM.lower():
+            # validate mhsm parameters prior to creating the encryption settings object
+            if not kv_url.strip() or not kv_id.strip():
+                raise Exception("KeyVaultUrl or KeyVaultResourceId are not empty. Please remove them for use with 'KeyStoreType':'ManagedHSM'.")
+            self.logger.log("get_settings_data: KeyVault Url is empty, validating KeyEncryptionKeyKVURL and KeyEncryptionKeyKVId for ManagedHSM")
+            cutil.check_mhsm_url(kek_url, "A ManagedHSM URL was specified, but it was invalid for ManagedHSM. ")
+            cutil.check_mhsm_id(kek_kv_id, "A ManagedHSM ID is required, but is missing or invalid. ")
+            cutil.check_mhsm_name(kek_kv_id, kek_url, "A ManagedHSM ID and ManagedHSM URL were provided, but their ManagedHSM names did not match")
+            self._DISK_ENCRYPTION_KEY_SOURCE = CommonVariables.KeyStoreTypeManagedHSM
+        elif key_store_type:
+            raise Exception("The expected flag name and value to enable ManagedHSM is 'KeyStoreType':'{0}'. " +
```

Comment:
Lets say the customer passes "`KeyStoreType":""` , in the template , So for this case the code doesn't raise the exception because it is empty and the condition `elif key_store_type:` returns `false`.

You should raise the exception when the key `KeyStoreType` exists in the `elif` case.

Reply:
Not sure I am following. If the customer passes "KeyStoreType":"" , correct, the exception would not be hit. It will only be hit if key_store_type exists and is not "ManagedHSM", so for example, "  " or "KeyVault" would throw. 

If the customer passes "KeyStoreType":"" with a ManagedHSM URL, I expect the KeyVault checks would fail. If we want it to fail automatically on this, I would need an "elif not key_store_type:" that throws an exception saying something like "Key Store Flag passed with null/empty value is not allowed, remove the flag or pass the value of 'ManagedHSM'"